### PR TITLE
Add property as IPC interface capability.

### PR DESCRIPTION
**Pull request by mamokarz**

### CODE REVIEW

File: inc/az_ulib_ipc_api.h [hunk 1]
```diff
@@ -360,6 +362,103 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+ * @return The #az_result with the result of the call.
+ *  @retval #AZ_OK                              If the IPC get success getting the property content.
+ *  @retval #AZ_ERROR_ITEM_NOT_FOUND            If the target property does not exist.
+ *  @retval #AZ_ERROR_NOT_SUPPORTED             If the target property does not support get with
+ *                                              string.
+ *  @retval Others                              Defined by the target `get` function.
+ */
+AZ_NODISCARD az_result az_ulib_ipc_get_with_str(
+    az_ulib_ipc_interface_handle interface_handle,
+    az_ulib_capability_index command_index,
```
Comment (mamokarz) [inline inc/az_ulib_ipc_api.h:411]: >command_index [](http://example.com/codeflow?start=29&length=13)

property_index

File: inc/az_ulib_ipc_api.h [hunk 2]
```diff
@@ -360,6 +362,103 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+ * @return The #az_result with the result of the call.
+ *  @retval #AZ_OK                              If the IPC get success setting the property content.
+ *  @retval #AZ_ERROR_ITEM_NOT_FOUND            If the target property does not exist.
+ *  @retval #AZ_ERROR_NOT_SUPPORTED             If the target property does not support set with
+ *                                              string.
+ *  @retval Others                              Defined by the target `get` function.
+ */
+AZ_NODISCARD az_result az_ulib_ipc_set_with_str(
+    az_ulib_ipc_interface_handle interface_handle,
+    az_ulib_capability_index command_index,
```
Comment (mamokarz) [inline inc/az_ulib_ipc_api.h:459]: >command_index [](http://example.com/codeflow?start=29&length=13)

property_index

**Comment by smcd253** [comment]
lgtm. make sure the command_index/property_index problem on clang and osx is addressed before merging.

File: inc/az_ulib_descriptor_api.h [hunk 3]
```diff
@@ -229,15 +230,17 @@ typedef void* az_ulib_ipc_interface_handle;
...
-    ._internal                                                              \
-        = {.name = AZ_SPAN_LITERAL_FROM_STR(command_name),                  \
-           .capability_ptr_1 = { .command_async = command_concrete },       \
-           .capability_ptr_2 = { .cancel = cancel_concrete },               \
-           .span_wrapper_ptr_1 = { .command_async = command_span_wrapper }, \
-           .flags = (uint8_t)(AZ_ULIB_CAPABILITY_TYPE_COMMAND_ASYNC) }      \
+#define AZ_ULIB_DESCRIPTOR_ADD_COMMAND_ASYNC(                                                     \
+    command_name, command_concrete, command_span_wrapper, cancel_concrete)                        \
+  {                                                                                               \
+    ._internal = {                                                                                \
```
Comment (cindydeng1998) [inline inc/az_ulib_descriptor_api.h:236]: nit: align braces on same line like in AZ_ULIB_DESCRIPTOR_ADD_COMMAND above, or give them separate lines in the two above

Comment (mamokarz) [reply inc/az_ulib_descriptor_api.h:236]: This alignend is defined by the clang-format as required by our code guidelines.

Comment (cindydeng1998) [reply inc/az_ulib_descriptor_api.h:236]: so it has to be different within the same file? see line 171 and 200 compared to 236/237

Comment (mamokarz) [reply inc/az_ulib_descriptor_api.h:236]: Yep, clang-format decides the "best" way. I've just sent you an email how to enable it in VSCode.

**Review by danewalton — APPROVED** [review]

File: inc/az_ulib_ipc_api.h [hunk 4]
```diff
@@ -360,6 +362,103 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+    az_ulib_capability_index property_index,
+    az_ulib_model_out model_out);
+
+/**
+ * @brief   Get content of a published property using string models.
+ *
+ * @param[in]   interface_handle    The #az_ulib_ipc_interface_handle with the interface handle.
+ *                                  It cannot be `NULL`. Call az_ulib_ipc_try_get_interface() to
+ *                                  get the interface handle.
+ * @param[in]   property_index      The #az_ulib_capability_index with the property index. Call
```
Comment (smcd253) [inline inc/az_ulib_ipc_api.h:394]: clang throws an error here. must be command_index.

**Review by smcd253 — APPROVED** [review]
lgtm

File: inc/az_ulib_ipc_api.h [hunk 5]
```diff
@@ -360,6 +362,103 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+ *
+ * @pre     IPC shall already be initialized.
+ * @pre     \p interface_handle shall not be 'NULL'.
+ *
+ * @return The #az_result with the result of the get.
+ *  @retval #AZ_OK                              If the IPC get success getting the property.
+ *  @retval #AZ_ERROR_ITEM_NOT_FOUND            If the target interface does not exist.
+ *  @retval Others                              Defined by the target get.
+ */
+AZ_NODISCARD az_result az_ulib_ipc_get(
```
Comment (mamokarz) [inline inc/az_ulib_ipc_api.h:383]: >az_ulib_ipc_get [](http://example.com/codeflow?start=23&length=15)

After some discussions offline, we decided that we don't need get and set APIs. So I'll remove those.

File: tests/inc/az_ulib_test_my_interface.h [hunk 6]
```diff
@@ -91,46 +93,12 @@ extern "C"
...
     az_ulib_ipc_interface_handle handle;
-    az_ulib_capability_index command_index;
+    az_ulib_capability_index capability_index;
     az_result return_result;
   } my_command_model_in;
 
   typedef az_result my_command_model_out;
 
-/*
- * Define model of my_command_async in my_interface.
```
Comment (smcd253) [inline tests/inc/az_ulib_test_my_interface.h:101]: I remember you saying that you were going to hold off on async capability until later. I see it was included earlier, is it still possible to use async commands with the generic capability interface?

**Review by smcd253 — APPROVED** [review]
lgtm

File: tests/inc/az_ulib_test_my_interface.h (hunk 6 above)
Comment (mamokarz) [reply tests/inc/az_ulib_test_my_interface.h:101]: I'll put it back whene we decide to implent it for real.

**Review by danewalton — APPROVED** [review]

**Review by cindydeng1998 — APPROVED** [review]
🌞

File: inc/az_ulib_ipc_interface.h [hunk 7]
```diff
@@ -55,137 +55,137 @@ typedef struct
...
 
-} az_ulib_ipc_vtable;
+} az_ulib_ipc_table;
 
 /*
  * @brief   Dynamically linked wrapper to az_ulib_ipc_publish().
  */
 AZ_INLINE AZ_NODISCARD az_result azi_ulib_ipc_publish(
-    const az_ulib_ipc_vtable* const vtable,
+    const az_ulib_ipc_table* const table,
```
[the remark this answers is not in the document]
Comment (mamokarz) [reply inc/az_ulib_ipc_interface.h:78]: It was a feedback from Jeffrey, vtable has a very specific meaning and it is not totally correct here.

**Pull request merged**